gcc/rust/expand/rust-macro-expand.cc

```diff
@@ -3811,53 +3861,63 @@ MacroExpander::transcribe_rule (
   //   as a statement (either via ExpressionStatement or
   //   MacroInvocationWithSemi)
 
-  // parse the item
-  std::vector<AST::SingleASTNode> nodes;
-  switch (invoc_token_tree.get_delim_type ())
+  // The flow-chart in order to choose a parsing function is as follows:
+  //
+  // [is in item context?]
+  //     -- Yes --> parser.parse_item();
+  //     -- No --> [has semicolon?]
+  //                 -- Yes --> parser.parse_stmt();
+  //                 -- No --> [switch invocation.delimiter()]
+  //                             -- { } --> parser.parse_stmt();
+  //                             -- _ --> parser.parse_expr();
+
+  // If there is a semicolon OR we are expanding a MacroInvocationSemi, then
+  // we can parse multiple items. Otherwise, parse multiple expressions
+  std::function<AST::SingleASTNode ()> parse_fn
+    = [&parser] () { return AST::SingleASTNode (parser.parse_expr ()); };
```

Comment:
great use of lambda's here. Though can we make this two statements:
```
std::unique_ptr<..> e = parser.parse_expr();
return AST::SingleASTNode(std::move(e));
```
This helps with gdb sessions